ee/management/commands/backfill_persons_and_groups_on_events.py

```diff
@@ -0,0 +1,148 @@
+import structlog
+from django.core.management.base import BaseCommand
+
+from ee.clickhouse.sql.events import EVENTS_DATA_TABLE
+from ee.clickhouse.sql.groups import GROUPS_TABLE
+from ee.clickhouse.sql.person import PERSON_DISTINCT_ID2_TABLE, PERSONS_TABLE
+from posthog.client import sync_execute
+from posthog.settings import CLICKHOUSE_CLUSTER, CLICKHOUSE_DATABASE
+
+"""
+WARNING: This script is in Alpha! Make sure you know what you're doing before running it with --live-run set.
+
+Pre-requisites:
+
+- `allow_nondeterministic_mutations` should be set to 1 on the current user's profile
+
+Approach:
+
+1. Create dictionaries for groups, persons, and person distinct ID tables
+
+These dictionaries will source data from the target Replicated tables, so they are guaranteed to be consistent across nodes.
+We keep a small cache in memory of 1000 frequently accessed keys for each dictionary. We hit the source table for values not in the cache.
+
+2. Alter table to update values according to the dictionaries
+
+With the dictionaries in place, we run an ALTER TABLE backfill. Note that this will have to rewrite a lot of data.
+This backfill will look up values in the dictionaries and set the corresponding columns to them.
+Also note that minor inconsitencies can occur from distinct IDs and properties changing during the backfill.
+
+3. Drop the dictionaries
+"""
+
+logger = structlog.get_logger(__name__)
+
+GROUPS_DICT_TABLE_NAME = f"{GROUPS_TABLE}_dict"
+PERSONS_DICT_TABLE_NAME = f"{PERSONS_TABLE}_dict"
+PERSON_DISTINCT_IDS_DICT_TABLE_NAME = f"{PERSON_DISTINCT_ID2_TABLE}_dict"
+
+GROUPS_DICTIONARY_SQL = f"""
+CREATE DICTIONARY IF NOT EXISTS {GROUPS_DICT_TABLE_NAME}
+ON CLUSTER '{CLICKHOUSE_CLUSTER}'
+(
+    group_key String,
+    group_properties String
+)
+PRIMARY KEY group_key
+SOURCE(CLICKHOUSE(TABLE {GROUPS_TABLE} DB '{CLICKHOUSE_DATABASE}' USER 'default'))
+LAYOUT(cache(size_in_cells 1000))
+Lifetime(60000)
+"""
+
+
+PERSON_DISTINCT_IDS_DICTIONARY_SQL = f"""
+CREATE DICTIONARY IF NOT EXISTS {PERSON_DISTINCT_IDS_DICT_TABLE_NAME}
+ON CLUSTER '{CLICKHOUSE_CLUSTER}'
+(
+    distinct_id String,
+    person_id UUID
+)
+PRIMARY KEY distinct_id
+SOURCE(CLICKHOUSE(TABLE person_distinct_id DB '{CLICKHOUSE_DATABASE}' USER 'default'))
+LAYOUT(cache(size_in_cells 1000))
+Lifetime(60000)
+"""
+
+PERSONS_DICTIONARY_SQL = f"""
+CREATE DICTIONARY IF NOT EXISTS {PERSONS_DICT_TABLE_NAME}
+ON CLUSTER '{CLICKHOUSE_CLUSTER}'
+(
+    id UUID,
+    properties String
+)
+PRIMARY KEY id
+SOURCE(CLICKHOUSE(TABLE person DB '{CLICKHOUSE_DATABASE}' USER 'default'))
+LAYOUT(cache(size_in_cells 1000))
+Lifetime(60000)
+"""
+
+BACKFILL_BASE_SQL = f"""
+ALTER TABLE {EVENTS_DATA_TABLE()}
+ON CLUSTER '{CLICKHOUSE_CLUSTER}'
```

Comment:
This is problematic: We run this on EVERY replica this way, meaning we rewrite and sync data $REPLICATION_FACTOR too many times. A better way is to run once per shard though I think we don't have the tooling for that.

Reply:
Note that this will specifically hit the backup nodes on cloud which in turn might hose these for I/O and cause an incident/disk outage.